Win32-network/src/System/Win32/Async/Socket/ByteString/Lazy.hs

```diff
@@ -0,0 +1,77 @@
+{-# LANGUAGE BangPatterns #-}
+
+module System.Win32.Async.Socket.ByteString.Lazy
+  ( send
+  , sendAll
+  ) where
+
+
+import           Control.Monad (when)
+import           Data.ByteString.Lazy (ByteString)
+import qualified Data.ByteString.Lazy as BL
+import           Data.ByteString.Unsafe (unsafeUseAsCStringLen)
+import           Data.Int (Int64)
+import           Foreign.C (CInt (..))
+import           Foreign.Marshal.Array (allocaArray)
+import           Foreign.Ptr
+import           Foreign.StablePtr (StablePtr)
+import           Foreign.Storable
+
+import           Network.Socket (Socket)
+import qualified Network.Socket as Socket
+
+import           System.Win32.Types (DWORD)
+
+import           System.Win32.Async.WSABuf
+import           System.Win32.Async.Internal
+
+
+-- | Sending each chunk using vectored I/O.  In one system call one can
+-- transmit at most 1024 chunks, it is safe to transmit less than 4194304
+-- bytes.
+--
+send :: Socket
+     -> ByteString
+     -> IO Int64
+send sock bs = do
+    let cs  = take maxNumChunks (BL.toChunks bs)
+        size = length cs
+    siz <- Socket.withFdSocket sock $ \fd -> allocaArray size $ \ptr ->
+             withPokes cs ptr $ \nwsabuf ->
+               wsaWaitForCompletion "send" (c_sendBuf fd ptr nwsabuf)
+    return $ fromIntegral siz
+  where
+    withPokes ss p f = loop ss p 0 0
+      where
+        loop (c:cs) q k !nwsabuf
+            | k < maxNumBytes = unsafeUseAsCStringLen c $ \(ptr, strlen) -> do
+                poke q $ WSABuf (fromIntegral strlen) (castPtr ptr)
+                loop cs
+                     (q `plusPtr` sizeOf (undefined :: WSABuf))
+                     (k + fromIntegral strlen)
+                     (nwsabuf + 1)
+            | otherwise = f nwsabuf
+        loop _ _ _ nwsabuf = f nwsabuf
+
+    maxNumBytes, maxNumChunks :: Int
+    maxNumBytes  = 4194304 -- maximum number of bytes to transmit in one system call
+    maxNumChunks = 1024    -- maximum number of chunks to transmit in one system call
+
+
+foreign import ccall safe "HsSendBuf"
+    c_sendBuf :: SOCKET
+              -> Ptr WSABuf  -- ^ lpBuffers
+              -> DWORD       -- ^ dwBufferCount
+              -> StablePtr b
+              -> IO ()
+
+
+sendAll :: Socket
+        -> ByteString
+        -> IO ()
+sendAll sock bs = do
```

Comment:
Network.Socket.Lazy.sendAll handles "sending" an empty bytestring? It is not clear to me that your version does.

Reply:
good catch, I'll fix this.

Reply:
Done.